icu4c/source/common/locid.cpp

```diff
@@ -2461,21 +2461,27 @@ Locale::setKeywordValue(const char* keywordName, const char* keywordValue, UErro
     int32_t newLength = uloc_setKeywordValue(keywordName, keywordValue, fullName,
                                              bufferLength, &status) + 1;
     /* Handle the case the current buffer is not enough to hold the new id */
-    if (status == U_BUFFER_OVERFLOW_ERROR) {
+    if (status == U_BUFFER_OVERFLOW_ERROR || status == U_STRING_NOT_TERMINATED_WARNING) {
         U_ASSERT(newLength > bufferLength);
         char* newFullName = (char *)uprv_malloc(newLength);
         if (newFullName == nullptr) {
             status = U_MEMORY_ALLOCATION_ERROR;
             return;
         }
-        uprv_strcpy(newFullName, fullName);
+        uprv_strncpy(newFullName, fullName, newLength-1);
+        if (status == U_STRING_NOT_TERMINATED_WARNING) {
+            status = U_ZERO_ERROR;
+            newFullName[newLength-1] = '\0';
+        }
         if (fullName != fullNameBuffer) {
             // if full Name is already on the heap, need to free it.
             uprv_free(fullName);
         }
         fullName = newFullName;
-        status = U_ZERO_ERROR;
-        uloc_setKeywordValue(keywordName, keywordValue, fullName, newLength, &status);
+        if (status == U_BUFFER_OVERFLOW_ERROR) {
+            status = U_ZERO_ERROR;
+            uloc_setKeywordValue(keywordName, keywordValue, fullName, newLength, &status);
+        }
     } else {
         U_ASSERT(newLength <= bufferLength);
```

Comment:
You can tighten this to `<` now.

Reply:
I don't think so the newLength is has a "+1" on top of what got returned by uloc_setKeywordValue in there already. (before my change)